Teleport occasionally missing end of session recordings

The following integration test intermittently fails which indicates that the end of a session is occasionally not recorded. This test fails very infrequently on Jenkins and our local developer workstations, but fails with some regularity on our macOS cloud box.

```
Xcloud@/tmp/teleport-tests/src/github.com/gravitational/teleport/integration$ GOPATH=/tmp/teleport-tests go test -v ./... -check.f=TestAuditOn
=== RUN   TestIntegrations
[AUTH]  Auth service is starting on localhost:20198
[Node] Successfully registered with the cluster
[Proxy] Successfully registered with the cluster
Starting reverse tunnel service is starting on localhost:20195 using no cache policy
[Node] Successfully registered with the cluster
[PROXY] SSH proxy service is starting on localhost:20197
[SSH]   Service is starting on localhost:20199 using no cache policy
[SSH]   Service is starting on localhost:20194 using cache policy with 20h0m0s TTL
the connection was closed on the remote side on  31 Jan 18 10:51 PST
[AUTH]  Auth service is starting on localhost:20192
[Node] Successfully registered with the cluster
[Proxy] Successfully registered with the cluster
Starting reverse tunnel service is starting on localhost:20189 using no cache policy
[PROXY] SSH proxy service is starting on localhost:20191
[SSH]   Service is starting on localhost:20193 using no cache policy
[Node] Successfully registered with the cluster
[SSH]   Service is starting on localhost:20188 using cache policy with 20h0m0s TTL
the connection was closed on the remote side on  31 Jan 18 10:51 PST

----------------------------------------------------------------------
FAIL: integration_test.go:161: IntSuite.TestAuditOn

integration_test.go:302:
    // session stream keeps coming back short
    c.Fatal("stream is not getting data: %q", string(sessionStream))
Xcloud@~$ echo hi
hi
Xcloud@~$ 

OOPS: 0 passed, 1 FAILED
--- FAIL: TestIntegrations (13.32s)
FAIL
exit status 1
FAIL	github.com/gravitational/teleport/integration	13.475s
```

--- a/lib/defaults/defaults.go
+++ b/lib/defaults/defaults.go
@@ -171,6 +171,10 @@ const (
 
 	// PlaybackRecycleTTL is the TTL for unpacked session playback files
 	PlaybackRecycleTTL = 3 * time.Hour
+
+	// WaitCopyTimeout is how long Teleport will wait for a session to finish
+	// copying data from the PTY after "exit-status" has been received.
+	WaitCopyTimeout = 5 * time.Second
 )
 
 var (

--- a/lib/srv/sess.go
+++ b/lib/srv/sess.go
@@ -614,18 +614,42 @@ func (s *session) start(ch ssh.Channel, ctx *ServerContext) error {
 	// the session server (terminal size and activity)
 	go s.pollAndSync()
 
-	// Pipe session to shell and visa-versa capturing input and output
+	doneCh := make(chan bool, 1)
+
+	// copy everything from the pty to the writer. this lets us capture all input
+	// and output of the session (because input is echoed to stdout in the pty).
+	// the writer contains multiple writers: the session logger and a direct
+	// connection to members of the "party" (other people in the session).
 	s.term.AddParty(1)
 	go func() {
-		// notify terminal about a copy process going on
 		defer s.term.AddParty(-1)
-		io.Copy(s.writer, s.term.PTY())
-		log.Infof("session.io.copy() stopped")
+
+		_, err := io.Copy(s.writer, s.term.PTY())
+		log.Debugf("Copying from PTY to writer completed with error %v.", err)
+
+		// once everything has been copied, notify the goroutine below. if this code
+		// is running in a teleport node, when the exec.Cmd is done it will close
+		// the PTY, allowing io.Copy to return. if this is a teleport forwarding
+		// node, when the remote side closes the channel (which is what s.term.PTY()
+		// returns) io.Copy will return.
+		doneCh <- true
 	}()
 
-	// wait for the shell to complete:
+	// wait for exec.Cmd (or receipt of "exit-status" for a forwarding node),
+	// once it is received wait for the io.Copy above to finish, then broadcast
+	// the "exit-status" to the client.
 	go func() {
 		result, err := s.term.Wait()
+
+		// wait for copying from the pty to be complete or a timeout before
+		// broadcasting the result (which will close the pty) if it has not been
+		// closed already.
+		select {
+		case <-time.After(defaults.WaitCopyTimeout):
+			log.Errorf("Timed out waiting for PTY copy to finish, session data for %v may be missing.", s.id)
+		case <-doneCh:
+		}
+
 		if result != nil {
 			s.registry.broadcastResult(s.id, *result)
 		}

